### PR TITLE
chore(ci): add numpy 2 release candidate to test matrix

### LINKED ISSUE #1818: NumPy 2.0 support
NumPy 2.0 is coming out soon ( https://github.com/numpy/numpy/issues/24300 ). NumPy 2.0.0rc1 packages for conda & wheels came out 3 weeks back ( https://github.com/numpy/numpy/issues/24300#issuecomment-2030603395 )

To prepare for NumPy 2.0, it might be worthwhile to start testing against NumPy 2 in CI

NumPy has put out [a migration guide]( https://numpy.org/devdocs/numpy_2_0_migration_guide.html#numpy-2-migration-guide ). More details are in [the release notes]( https://numpy.org/devdocs/release/2.0.0-notes.html )

Also as NumPy is tracking ecosystem support for NumPy 2.0, it would be helpful to share Zarr's current support status in issue (with any plans): https://github.com/numpy/numpy/issues/26191

### PR DESCRIPTION
**Pull request by jhamman**
closes #1818 

cc @jakirkham 

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by jakirkham** [comment]
Thanks Joe! 🙏

Looks good

Was trying to look in the CI logs to see where NumPy gets installed, but was having trouble seeing it. Do you know where this happens?

**Comment by jhamman** [comment]
@jakirkham - your comment uncovered something in our new CI setup. We weren't passing the numpy version to the list of dependencies. That is fixed in this PR now. 

Note: the test failures are in the v2 code path. 

@maxrjones - could you take a look here and tell us if the `features` / `version` part of the matrix envs are getting picked up correctly?

File: pyproject.toml [hunk 1]
```diff
@@ -107,19 +110,20 @@ features = ["extra"]
...
-numpy = ["1.24", "1.26"]
+numpy = ["1.24", "1.26", "2.0.0rc1"]
 features = ["optional"]
 
 [tool.hatch.envs.test.scripts]
 run-coverage = "pytest --cov-config=pyproject.toml --cov=pkg --cov=tests"
 run = "run-coverage --no-cov"
 run-verbose = "run-coverage --verbose"
 run-mypy = "mypy src"
+list-env = "pip list"
```
Comment (jhamman) [inline pyproject.toml:126]: this is used in the test github action

File: pyproject.toml [hunk 2]
```diff
@@ -92,6 +92,9 @@ version.source = "vcs"
 build.hooks.vcs.version-file = "src/zarr/_version.py"
 
 [tool.hatch.envs.test]
+dependencies = [
+    "numpy~={matrix:numpy}",
+]
```
Comment (jhamman) [inline pyproject.toml:97]: this was what we were missing. Eeek!

**Comment by jakirkham** [comment]
> your comment uncovered something in our new CI setup. We weren't passing the numpy version to the list of dependencies. That is fixed in this PR now.

Sounds like a productive exercise then 🙂

Thanks Joe! 🙏

**Comment by maxrjones** [comment]
> @maxrjones - could you take a look here and tell us if the features / version part of the matrix envs are getting picked up correctly?

The only dependency that is different between `minimal` and `optional` is `lmdb` which is also included in the extra-dependencies for the tests env, meaning the features part of the matrix is not currently doing anything. If you're willing to clarify which deps should be including in only optional I can fix that part of the setup.

**Comment by jakirkham** [comment]
Looking good. Thanks Joe! 🙏

Can see NumPy 2.0.0rc1 getting installed and CI passing 🎉

**Review by jakirkham — APPROVED** [review]

**Pull request merged**